### DISCUSSION TITLE
Unique container per macOS window

### DISCUSSION THREAD

**Discussion by NickRussell95** [hmlongco/Factory#229] [Q&A] [answered]
Hello 👋 

I'm wondering if there is a recommended pattern to create a `Container` per window in a SwiftUI macOS app that contains a `WindowGroup`.

Basically, I'm storing some global application state in my services, but I would like this state to be unique per `Window` within the `WindowGroup`. 

eg. this is very simplified example. But the problem is that `currentTime` in `TimeService` is shared across all active windows. I would like this to be unique per window.

```Swift
@Observable class TimeService {
    var currentTime = Date(timeIntervalSince1970: 0)
}
```
```Swift
extension Container {
    var timeService: Factory<TimeService> {
        Factory(self) { TimeService() }
    }
}
```
```Swift
@Observable class ClockViewModel {
    
    @ObservationIgnored
    @Injected(\.timeService) var timeService
    
    var time: Double {
        get {
            timeService.currentTime.timeIntervalSince1970
        }
        set {
            timeService.currentTime = Date(timeIntervalSince1970: newValue)
        }
    }
}
```
```Swift

struct ClockView: View {
    @State var viewModel = ClockViewModel()
    
    var body: some View {
        Slider(value: $viewModel.time, in: 0...120)
    }
}
```
```Swift
@main
struct ExampleApp: App {
    var body: some Scene {
        WindowGroup {
            ClockView()
        }
    }
}

```

Thank you for any help!

**Comment by hmlongco** [comment, marked as answer]
You can pass a container along in view initializers, or via the Environment, but if you use any of the injected property wrappers those all interact with the global shared state for Container. There's no differentiation per scene.
